matrix_sdk_base/src/store/sled_store/mod.rs

```diff
@@ -1135,4 +1604,551 @@ mod test {
+
+        // Add an overlapping timeline slice to the store
+        let messages = MessageResponse::try_from_http_response(
+            Response::builder()
+                .body(serde_json::to_vec(&*test_json::OVERLAPPING_ROOM_MESSAGES_BATCH_1).unwrap())
+                .unwrap(),
+        )
+        .unwrap();
+
+        let events: Vec<SyncRoomEvent> = messages.chunk.iter().cloned().map(Into::into).collect();
+
+        stored_events.push(events.last().cloned().unwrap());
+
+        let timeline_slice = TimelineSlice::new(
+            events.clone(),
+            messages.start.clone().unwrap(),
+            messages.end.clone(),
+        );
+        let mut changes = StateChanges::default();
+        changes.add_timeline(&room_id, timeline_slice);
+        store.save_changes(&changes).await.unwrap();
+
+        // Read all of the known timeline from the beginning backwards
+        let backward_events = store
+            .get_timeline(&room_id, None, None, None, Direction::Backward)
+            .await
+            .unwrap()
+            .unwrap();
+
+        let end_token = messages.end;
+        assert_eq!(backward_events.events.len(), stored_events.len());
+        assert_eq!(backward_events.token, end_token);
+        assert!(!backward_events
+            .events
+            .iter()
+            .zip(stored_events.iter())
+            .any(|(a, b)| a.event_id() != b.event_id()));
+
+        // Add an overlapping patch to the start of everything
+        let messages = MessageResponse::try_from_http_response(
+            Response::builder()
+                .body(serde_json::to_vec(&*test_json::OVERLAPPING_ROOM_MESSAGES_BATCH_2).unwrap())
+                .unwrap(),
+        )
+        .unwrap();
+
+        let events: Vec<SyncRoomEvent> = messages.chunk.iter().cloned().map(Into::into).collect();
+
+        let mut expected_events = events.clone();
+        expected_events.extend_from_slice(&stored_events[1..]);
+
+        let timeline_slice = TimelineSlice::new(
+            events.clone(),
+            messages.start.clone().unwrap(),
+            messages.end.clone(),
+        );
+        let mut changes = StateChanges::default();
+        changes.add_timeline(&room_id, timeline_slice);
+        store.save_changes(&changes).await.unwrap();
+
+        let start_event = event_id!("$1444812213350496Cbbbr3:example.com");
+        // Read all of the known timeline from the beginning backwards
+        let backward_events = store
+            .get_timeline(&room_id, Some(&start_event), None, None, Direction::Backward)
+            .await
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(backward_events.events.len(), expected_events.len());
+        assert_eq!(backward_events.token, end_token);
+        assert!(!backward_events
+            .events
+            .iter()
+            .zip(expected_events.iter())
+            .any(|(a, b)| a.event_id() != b.event_id()));
```

Comment:
```suggestion
        assert!(backward_events
            .events
            .iter()
            .zip(expected_events.iter())
            .all(|(a, b)| a.event_id() == b.event_id()))
```

The double negative was confusing it took me a bit to figure out what was going on haha. Same for all of the above `! !=`

Reply:
i didn't see `all` makes much more sense :)